### PR TITLE
bpf: nat: improve logic that creates the NAT entries

### PR DESCRIPTION
**Pull request by julianwiedmann**
The logic in `snat_v*_new_mapping()` is super tricky, and I start afresh every time I need to look at it. Re-organize the code, and add comments to the crucial pieces.

### CODE REVIEW

**Comment by julianwiedmann** [comment]
Looks like this also nicely reduces complexity in the relevant code paths (from pulling the NAT entry creation out of the loop?).

File: bpf/lib/nat.h [hunk 1]
```diff
@@ -221,7 +220,13 @@ static __always_inline int snat_v4_new_mapping(struct __ctx_buff *ctx,
 
 #pragma unroll
 	for (retries = 0; retries < SNAT_COLLISION_RETRIES; retries++) {
+		rtuple.dport = bpf_htons(port);
+
+		/* Check if the selected port is already in use by a RevSNAT
+		 * entry for some other connection with the same src/dst:
+		 */
 		if (!__snat_lookup(map, &rtuple)) {
+			ostate->to_sport = rtuple.dport;
```
Comment (gentoo-root) [inline bpf/lib/nat.h]: >Also only update the SNAT tuple if the check was successful and the
>currently selected port is available.

__snat_update can still fail, and ostate will be changed, but I guess it's fine, given that we used to update it unconditionally.

File: bpf/lib/nat.h [hunk 2]
```diff
@@ -171,28 +171,15 @@ struct ipv4_nat_entry *snat_v4_lookup(const struct ipv4_ct_tuple *tuple)
 	return __snat_lookup(&SNAT_MAPPING_IPV4, tuple);
 }
```
Comment (gentoo-root) [inline bpf/lib/nat.h:173]: >remove *swat_tuple() helpers

Typo in the commit title.

**Review by gentoo-root — APPROVED** [review]
Looks good, no bugs spotted.

File: bpf/lib/nat.h (hunk 1 above)
Comment (julianwiedmann) [reply bpf/lib/nat.h]: Yep, I think reverting the change in that case would be excessive (given that the caller should be dropping the packet on such an error, not looking at the SNAT entry).

I should fix up the patch description though:
>Also only update the SNAT ~~tuple~~ entry if the check was successful and the
currently selected port is available.

**Comment by julianwiedmann** [comment]
No-change push to fix up the patch descriptions.

**Pull request merged**